src/tools/fuzzing/heap-types.cpp

```diff
@@ -95,32 +103,169 @@ struct HeapTypeGeneratorImpl {
+    // remaining descriptors we have committed to create in this group.
+    std::vector<Index> describees;
+    size_t committed = 0;
+
     size_t end = start + size;
     for (size_t i = start; i < end; ++i) {
       recGroupEnds.push_back(end);
-      planType(i, numRoots);
+      planType(i, numRoots, end - i, describees, committed);
     }
     return size;
   }
 
-  void planType(size_t i, size_t numRoots) {
+  void planType(size_t i,
+                size_t numRoots,
+                size_t remaining,
+                std::vector<Index>& describees,
+                size_t& committed) {
+    assert(remaining >= committed);
     typeIndices.insert({builder[i], i});
     // Everything is a subtype of itself.
     subtypeIndices[i].push_back(i);
-    if (i < numRoots || rand.oneIn(2)) {
+
+    // We may pick a supertype. If we have a described type that itself has a
+    // supertype, then we must choose that supertype's descriptor as our
+    // supertype.
+    std::optional<Index> super;
+
+    // Pick a type to describe, or choose not to describe a type by
+    // picking the one-past-the-end index. If all of the remaining types must be
+    // descriptors, then we must choose a describee.
+    Index describee = rand.upTo(describees.size() + (remaining != committed));
+
+    bool isDescriptor = false;
+    if (describee != describees.size()) {
+      isDescriptor = true;
+      --committed;
+
+      // If the intended described type has a supertype with a descriptor, then
+      // that descriptor must be the supertype of the type we intend to
+      // generate. However, we may not have generated that descriptor yet,
+      // meaning it is unavailable to be the supertype of the current type.
+      // Detect that situation and plan to generate the missing supertype
+      // instead.
+      Index described;
+      while (true) {
+        described = describees[describee];
+        auto describedSuper = supertypeIndices[described];
+        if (!describedSuper) {
+          // The described type has no supertype, so there is no problem.
+          break;
+        }
+        if (descriptorChainLengths[*describedSuper] == 0) {
+          // The supertype of the described type will not have a descriptor,
+          // so there is no problem.
+          break;
+        }
+        if ((super = descriptorIndices[*describedSuper])) {
+          // The descriptor of the described type's supertype, which must become
+          // the current type's supertype, has already been generated. There is
+          // no problem.
+          break;
+        }
+        // The necessary supertype does not yet exist. Find its described type
+        // so we can try to generate the missing supertype instead.
+        for (describee = 0; describee < describees.size(); ++describee) {
+          if (describees[describee] == *describedSuper) {
+            break;
+          }
+        }
+        assert(describee < describees.size());
```

Comment:
This assert could be at the top of the loop, before `described = describees[describee];`